File: tests/storage/set_obdata.c
/* Copyright (C) 2013, 2014 Roland Lutz

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software Foundation,
   Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.  */

#include <xornstorage.h>
#include <assert.h>
#include <stdlib.h>
#include <string.h>


static void check(xorn_revision_t rev, xorn_object_t ob0, xorn_object_t ob1)
{
	xorn_object_t *objects;
	size_t count;

	assert(xorn_get_objects(rev, &objects, &count) == 0);
	assert(objects != NULL);
	assert(count == 2);
	assert(objects[0] == ob0);
	assert(objects[1] == ob1);
	free(objects);
}

int main()
{
	xorn_revision_t rev;
	xorn_object_t ob0, ob1;
	struct xornsch_line line_data;
	struct xornsch_net net_data;
	struct xornsch_component component_data;
	struct xornsch_text text_data;

	memset(&line_data, 0, sizeof line_data);
	memset(&net_data, 0, sizeof net_data);
	memset(&component_data, 0, sizeof component_data);
	memset(&text_data, 0, sizeof text_data);

	assert(rev = xorn_new_revision(NULL));
	assert(ob0 = xornsch_add_line(rev, &line_data));
	assert(ob1 = xornsch_add_line(rev, &line_data));

	assert(xornsch_set_net_data(rev, ob0, &net_data) == 0);
	assert(xornsch_set_text_data(rev, ob1, &text_data) == 0);

	check(rev, ob0, ob1);


	/* data can be set without the object being reordered */

	assert(xornsch_set_line_data(rev, ob0, &line_data) == 0);
	check(rev, ob0, ob1);
	assert(xornsch_set_net_data(rev, ob0, &net_data) == 0);
	check(rev, ob0, ob1);

	assert(xornsch_set_line_data(rev, ob1, &line_data) == 0);
	check(rev, ob0, ob1);
	assert(xornsch_set_net_data(rev, ob1, &net_data) == 0);
	check(rev, ob0, ob1);

	/* objects are re-added at the end */

	xorn_delete_object(rev, ob0);
	assert(xornsch_set_line_data(rev, ob0, &line_data) == 0);
	check(rev, ob1, ob0);
	xorn_delete_object(rev, ob0);
	assert(xornsch_set_line_data(rev, ob0, &line_data) == 0);
	check(rev, ob1, ob0);

	xorn_delete_object(rev, ob1);
	assert(xornsch_set_line_data(rev, ob1, &line_data) == 0);
	check(rev, ob0, ob1);
	xorn_delete_object(rev, ob1);
	assert(xornsch_set_line_data(rev, ob1, &line_data) == 0);
	check(rev, ob0, ob1);

	xorn_delete_object(rev, ob0);
	assert(xornsch_set_net_data(rev, ob0, &net_data) == 0);
	check(rev, ob1, ob0);
	xorn_delete_object(rev, ob0);
	assert(xornsch_set_net_data(rev, ob0, &net_data) == 0);
	check(rev, ob1, ob0);

	xorn_delete_object(rev, ob1);
	assert(xornsch_set_net_data(rev, ob1, &net_data) == 0);
	check(rev, ob0, ob1);
	xorn_delete_object(rev, ob1);
	assert(xornsch_set_net_data(rev, ob1, &net_data) == 0);
	check(rev, ob0, ob1);


	xorn_free_revision(rev);
	return 0;
}

File: tests/cpython/set_obdata.py
# Copyright (C) 2013-2016 Roland Lutz
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software Foundation,
# Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.

import xorn.storage

def check_not_attached(rev, ob0, ob1):
    assert rev.get_objects() == [ob0, ob1]
    rev.relocate_object(ob0, None, None)
    assert rev.get_objects() == [ob1, ob0]
    rev.relocate_object(ob0, None, ob1)

def check_attached(rev, ob0, ob1):
    assert rev.get_objects() == [ob0, ob1]
    rev.relocate_object(ob0, None, None)
    assert rev.get_objects() == [ob0, ob1]

def set_object_data_fails(rev, ob, data):
    try:
        rev.set_object_data(ob, data)
    except ValueError:
        return True
    else:
        return False

line_data = xorn.storage.Line()
net_data = xorn.storage.Net()
component_data = xorn.storage.Component()
text_data = xorn.storage.Text()

rev = xorn.storage.Revision()
ob0 = rev.add_object(line_data)
ob1 = rev.add_object(line_data)

assert rev is not None
assert ob0 is not None
assert ob1 is not None

rev.set_object_data(ob0, net_data)
rev.set_object_data(ob1, text_data)

check_not_attached(rev, ob0, ob1)


# data can be set without the object being reordered

rev.set_object_data(ob0, line_data)
check_not_attached(rev, ob0, ob1)
rev.set_object_data(ob0, net_data)
check_not_attached(rev, ob0, ob1)

rev.set_object_data(ob1, line_data)
check_not_attached(rev, ob0, ob1)
rev.set_object_data(ob1, net_data)
check_not_attached(rev, ob0, ob1)

# objects are re-added at the end

rev.delete_object(ob0)
rev.set_object_data(ob0, line_data)
check_not_attached(rev, ob1, ob0)
rev.delete_object(ob0)
rev.set_object_data(ob0, line_data)
check_not_attached(rev, ob1, ob0)

rev.delete_object(ob1)
rev.set_object_data(ob1, line_data)
check_not_attached(rev, ob0, ob1)
rev.delete_object(ob1)
rev.set_object_data(ob1, line_data)
check_not_attached(rev, ob0, ob1)

rev.delete_object(ob0)
rev.set_object_data(ob0, net_data)
check_not_attached(rev, ob1, ob0)
rev.delete_object(ob0)
rev.set_object_data(ob0, net_data)
check_not_attached(rev, ob1, ob0)

rev.delete_object(ob1)
rev.set_object_data(ob1, net_data)
check_not_attached(rev, ob0, ob1)
rev.delete_object(ob1)
rev.set_object_data(ob1, net_data)
check_not_attached(rev, ob0, ob1)


# can change ob0 to any type while no object is attached

rev.set_object_data(ob0, line_data)
check_not_attached(rev, ob0, ob1)
rev.set_object_data(ob0, component_data)
check_not_attached(rev, ob0, ob1)
rev.set_object_data(ob0, text_data)
check_not_attached(rev, ob0, ob1)
rev.set_object_data(ob0, net_data)
check_not_attached(rev, ob0, ob1)

# can change type of ob1 while not attached

rev.set_object_data(ob1, line_data)
check_not_attached(rev, ob0, ob1)
rev.set_object_data(ob1, net_data)
check_not_attached(rev, ob0, ob1)
rev.set_object_data(ob1, component_data)
check_not_attached(rev, ob0, ob1)
rev.set_object_data(ob1, text_data)
check_not_attached(rev, ob0, ob1)


rev.relocate_object(ob1, ob0, None)
check_attached(rev, ob0, ob1)


# can't change ob0 to line or text while an object is attached

assert set_object_data_fails(rev, ob0, line_data)
check_attached(rev, ob0, ob1)
rev.set_object_data(ob0, net_data)
check_attached(rev, ob0, ob1)
rev.set_object_data(ob0, component_data)
check_attached(rev, ob0, ob1)
assert set_object_data_fails(rev, ob0, text_data)
check_attached(rev, ob0, ob1)

# can't change type of ob1 while attached

assert set_object_data_fails(rev, ob1, line_data)
check_attached(rev, ob0, ob1)
assert set_object_data_fails(rev, ob1, net_data)
check_attached(rev, ob0, ob1)
assert set_object_data_fails(rev, ob1, component_data)
check_attached(rev, ob0, ob1)
rev.set_object_data(ob1, text_data)
check_attached(rev, ob0, ob1)
